src/shared.rs

```diff
@@ -64,39 +70,51 @@ where
 }
 
 /// Read the next null-delimited JSON instruction from a stream.
-pub async fn recv_json<T: DeserializeOwned>(
-    reader: &mut (impl AsyncBufRead + Unpin),
-    buf: &mut Vec<u8>,
+pub async fn recv_json<T: DeserializeOwned, U: AsyncRead + AsyncWrite + Unpin>(
+    reader: &mut Framed<U, AnyDelimiterCodec>,
 ) -> Result<Option<T>> {
     trace!("waiting to receive json message");
-    buf.clear();
-    reader.read_until(0, buf).await?;
-    if buf.is_empty() {
-        return Ok(None);
-    }
-    if buf.last() == Some(&0) {
-        buf.pop();
+    if let Some(frame) = reader.next().await {
+        Ok(frame
+            .map(|op| {
+                serde_json::from_slice(&op.to_vec())
+                    .context("unable to parse message")
+                    .unwrap()
+            })
+            .context("frame error, invalid chunck length")?)
```

Comment:
Use `?` instead of `map` and don't use `unwrap` since that will crash the program. In particular it would allow anyone to crash the server on an invalid slice, which is a much larger security issue.